nagios monitor with event for context (#1553)
signalfx/signalfx-agent: docs/monitors/nagios.md

Monitor Type: `nagios` ([Source](https://github.com/signalfx/signalfx-agent/tree/master/pkg/monitors/nagios))

**Accepts Endpoints**: No

**Multiple Instances Allowed**: Yes

## Overview

Wrapper to run existing nagios status check scripts through SignalFx agent which will play the 
role of NRPE or SNMP exec.

It will run the script set in `command` parameter and send the 
[state](https://nagios-plugins.org/doc/guidelines.html#AEN78) of the check depending on the 
exit code of the command.

It is very similar to [telegraf/exec](https://docs.signalfx.com/en/latest/integrations/agent/monitors/telegraf-exec.html) 
monitor configured with `dataFormat: nagios` but:
  - it does not retrieve perfdata metrics, only the state of the script for alerting purpose.
  - it will override the state if exit code == 0 but output string starts with `warn`, `crit` or `unkn` 
  (not case sensitive).

Also the main advantage and purpose of this monitor is to add more context to this status check state 
thougth SignalFx [events](https://docs.signalfx.com/en/latest/detect-alert/events-intro.html).
Indeed, in addition to the state metric, it will send an event which includes the output and the error 

> catched from the command execution.

Reviewer: ```suggestion
caught from the command execution.
```

> This should make the troubleshooting more efficient and allow the user to remain in SignalFx withtout 

Reviewer: ```suggestion
This should make the troubleshooting more efficient and allow the user to remain in SignalFx without 
```

> to have to connect to the machine in case of anormal state to understand what is happen.

Reviewer: ```suggestion
to have to connect to the machine in case of anormal state to understand what is happening.
```

> It will also give the ability to create a dashboard similar to what nagios user are used to have.

Reviewer: ```suggestion
It will also give the ability to create a dashboard similar to what nagios user are accustomed to.
```

__Note__: the last sent event is cached into memory to avoid sending the same event for each collection 
interval over and over but already sent event will be send again when you restart the agent erasing its 
cache.

## Configuration

To activate this monitor in the Smart Agent, add the following to your
agent config:

```
monitors:  # All monitor config goes under this key
 - type: nagios
   ...  # Additional config
```

**For a list of monitor options that are common to all monitors, see [Common
Configuration](../monitor-config.md#common-configuration).**

| Config option | Required | Type | Description |
| --- | --- | --- | --- |
| `command` | **yes** | `string` | The command to exec with any arguments like: `"LC_ALL=\"en_US.utf8\" /usr/lib/nagios/plugins/check_ntp_time -H pool.ntp.typhon.net -w 0.5 -c 1"` |
| `service` | **yes** | `string` | Corresponds to the nagios `service` column and allows to aggregate all instances of the same service (when calling the same check script with different arguments) |

> | `timeout` | no | `integer` | The max execution time allowed in seconds be fore to send SIGTERM. In any case, a SIGKILL will be send at `intervalSeconds` value. (**default:** `9`) |

Reviewer: ```suggestion
| `timeout` | no | `integer` | The max execution time allowed in seconds before sending SIGTERM. In any case, a SIGKILL will be sent at `intervalSeconds` value. (**default:** `9`) |
```
